### DISCUSSION TITLE
Only partial metrics sent to statsd when the airflow services run on different hosts (AWS Fargate tasks)

### DISCUSSION THREAD

**Discussion by nicolamarangoni** [apache/airflow#27393] [General]
### Apache Airflow version

2.4.1

### What happened

We have an AirFlow-Deployment on **AWS Fargate** where each service (webserver, scheduler, dag-processor, celery-workers) are running in their own task, each task has a different hostname and ip.
We also have a separated task with a **datadog-agent** running a statsd (**dogstatsd**) server to which we want to push airflow metrics.
We have `statsd_host=DATADOG_AGENT_HOST` and `statsd_datadog_enabled="True"`.
Statsd successfully collect metrics like **dagbag_size** and **scheduler_heartbeat** but **NOT** **healthy** and **can_connect**.
We similar deployment on k8s in which all service containers run in the same pod, also with a single hostname and ip for everything. In some k8s setups we have the datadog-agent in the same pod in another pod. In the second case we set **statsd_host = DDAGENT_POD_HOST** and we successfully collect all metrics.

My suspect is that metrics are not pushed to statsd from every service independently but **only by the scheduler service**.
The scheduler service check **health** and **can_connect** and push them to statsd. However these 2 metrics originates in the webserver. If the webserver runs in a different host than the scheduler (like different tasks in Fargate), the metrics are not collected.

Is it possible that the scheduler is the only service that push metrics to statsd and that it can collect metrics from other services always searching for them on localhost?

### What you think should happen instead

All metrics should be sent to statsd.

### How to reproduce

Setup AirFlow with every service running in a separated pod/host/task with its own ip and hostname.
Activate statsd.

### Operating System

Debian GNU/Linux 11 (bullseye)

### Versions of Apache Airflow Providers

apache-airflow-providers-amazon==6.0.0
apache-airflow-providers-celery==3.0.0
apache-airflow-providers-cncf-kubernetes==4.4.0
apache-airflow-providers-common-sql==1.2.0
apache-airflow-providers-databricks==3.3.0
apache-airflow-providers-datadog==3.0.0
apache-airflow-providers-dbt-cloud==2.2.0
apache-airflow-providers-docker==3.2.0
apache-airflow-providers-elasticsearch==4.2.1
apache-airflow-providers-ftp==3.1.0
apache-airflow-providers-google==8.3.0
apache-airflow-providers-grpc==3.0.0
apache-airflow-providers-hashicorp==3.1.0
apache-airflow-providers-http==4.0.0
apache-airflow-providers-imap==3.0.0
apache-airflow-providers-microsoft-azure==4.3.0
apache-airflow-providers-mysql==3.2.1
apache-airflow-providers-odbc==3.1.2
apache-airflow-providers-postgres==5.2.2
apache-airflow-providers-redis==3.0.0
apache-airflow-providers-sendgrid==3.0.0
apache-airflow-providers-sftp==4.1.0
apache-airflow-providers-slack==5.1.0
apache-airflow-providers-snowflake==3.3.0
apache-airflow-providers-sqlite==3.2.1
apache-airflow-providers-ssh==3.2.0

### Deployment

Other

### Deployment details

AirFlow on AWS Fargate (ECS) with separated tasks for every service.

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by potiuk** [comment]
Converted to a discussion. It's likely misunderstanding and misalignment of expectations.

I am not sure what you mean by "heathy" and "can_connect" metrics are and where you get the idea where it is sent by the webserver (for sure no information from webserver is sent through scheduler). Basically scheduler is not even aware of Webserver existence - webserver is merely client to webserver.

From what I understand - Airflow uses statsd to send "application" metrics not "healthiness" or "can_connect" metrics. Those should be collected differently - you can for example see how it is done in our HelmChart - there you can see liveness Probes and the like. Generally assessing healthiness and connectability of Airlfow components is a deployment thing, not airflow's - our Chart is a good example of that so if you are interested in those you should do it similarly and use statsd only to gather applicaiton-level (logical) metrics.

**Comment by nicolamarangoni** [comment]
We have a DD-Agent sidecar-container in the pods on our on-prem kubernetes, and it send, among others, the metrics "can_connect" and "healthy" to DataDog.
Running a similar setup with a DD-Agent container in AWS Fargate, I'm missing "can_connect" and "healthy", but I see "dagbag_size" and "scheduler_heartbeat".
The only difference I see between the 2 setups, is that in kubernetes, all containers run in the same pod with the same hostname and can address services in other containers using localhost:port. On the other hand, the fargate setup has containers running in separated hosts with separated ips.
As I don't know exactly how metrics are collected by airflow and sent to statsd, I just supposed that only the scheduler sends metrics to statsd, eventually collecting some of them from the webserver running on the same host.
According to my assumption, if scheduler and webserver run on the same host. The scheduler collect "healthy" and "can_connect" from, for example, localhost:8080, and push them to the statsd service running in the DD-agent.
It works on kubernetes, but not on AWS fargate, as the the scheduler doesn't find the webserver over localhost.

I cannot verify these assumptions as I don't find any documents regarding this.

The official chart is not really relevant because in our case we use DataDog, and the statsd server running in the datadog-agent.

**Reply by potiuk** [reply]
You can see all the documentation here https://airflow.apache.org/docs/apache-airflow/stable/logging-monitoring/metrics.html  (and there for example you will find that you can selectively disable the metrics for each component).

And this is open-source project - you can always check the code if in doubt, also if you want to track down what's going on you can modify and experiment with it.

Relevant code here: https://github.com/apache/airflow/blob/main/airflow/stats.py

Your assumption is, unfortunately, wrong and you need to dig deeper. 

There is no magic or special behaviour here. Airflow components - each one of them - sends statsd metrics using statsd client of your choice - directly to the statds server that handles it. How you add your statds servers is up to your deployment, None of the Airlfow components is a router, none of them serve as a router. You can even (if you look at the docs) configure the statsd logger.

Likely your problem is the way how DD-Agent uses certain metrics, processes or whateve to generate the "can_connect" and "healthy" metrics and I think you should start from understanding where DD-Agent takes it from (I have personally no idea).

**Reply by potiuk** [reply]
Also - you can always use the official Helm Chart as a reference to see how we are doing it. Even if you are not using it, it addresses a number of non-obvious problems you might have and looking at it might be very good way for you to understand how to fix your problems.

There is a good reason we have Helm Chart of ours - as deploying Airflow has a lot of moving parts, and even if you decided to not follow it and go your own path, it's always a good idea to learn how the community handled some of that - "standing on the shoulders of giants" as the saying is.

**Comment by eugeneYWang** [comment]
I can answer this question as our setup only have metrics `can_connect` and `healthy`. Maybe you already solve it but for future readers: these two metrics seems to be coming from probing requests from Datadog Agent to Airflow web server health endpoint. If you have properly setup `/etc/datadog-agent/conf.d/airflow.d/conf.yaml` and give it accessible `url` of airflow web server (accessible from the perspective of Datadog Agent). Datadog can probe the airflow's health endpoint.

But I feel you. I would complains Datadog for their poor maintenance of documentation. https://github.com/DataDog/integrations-core/blob/master/airflow/README.md. 

Even now their doc still have nothing about `AIRFLOW__METRICS__STATSD_DATADOG_ENABLED`, and I cannot get all metrics of those by following their doc. Trying to catch attention of some Datadog staffs @hestonhoffman @FlorentClarret @yzhan289
